lib/dispatch.go

```diff
@@ -259,92 +274,118 @@ func (inst *Instance) registerOne(ourName string, methods MethodSet, impl interf
+			regFail("%s: third input param must be a pointer, got %v", funcName, mType)
 		}
 		mType = mType.Elem()
 		if mType != inType {
-			panic(fmt.Sprintf("%s: third input param must match impl, expect %v, got %v", funcName, inType, mType))
+			regFail("%s: third input param must match impl, expect %v, got %v", funcName, inType, mType)
 		}
 		// Validate the output values of the implementation
 		msetNumOuts := f.NumOut()
 		if msetNumOuts < 1 || msetNumOuts > 3 {
-			panic(fmt.Sprintf("%s: bad number of outputs: %d", funcName, f.NumOut()))
+			regFail("%s: bad number of outputs: %d", funcName, f.NumOut())
 		}
 		// First output, if there's more than 1, matches the impl output
 		if msetNumOuts == 2 || msetNumOuts == 3 {
 			mType = f.Out(0)
 			if mType != outType {
-				panic(fmt.Sprintf("%s: first output val must match impl, expect %v, got %v", funcName, outType, mType))
+				regFail("%s: first output val must match impl, expect %v, got %v", funcName, outType, mType)
 			}
 		}
 		// Second output, if there are three, must be a cursor
 		if msetNumOuts == 3 {
 			mType = f.Out(1)
 			if mType.Name() != "Cursor" {
-				panic(fmt.Sprintf("%s: second output val must match a cursor, got %v", funcName, mType))
+				regFail("%s: second output val must match a cursor, got %v", funcName, mType)
 			}
 		}
 		// Last output must be an error
 		mType = f.Out(msetNumOuts - 1)
 		if mType.Name() != "error" {
-			panic(fmt.Sprintf("%s: last output val should be error, got %v", funcName, mType))
+			regFail("%s: last output val should be error, got %v", funcName, mType)
 		}
 
 		// Remove this method from the methodSetMap now that it has been processed
 		delete(methodMap, i.Name)
 
+		var endpoint APIEndpoint
+		var httpVerb string
+		// Additional attributes for the method are found in the Attributes
+		amap := methods.Attributes()
+		methodAttrs, ok := amap[lowerName]
+		if !ok {
+			regFail("not in Attributes: %s.%s", ourName, lowerName)
+		}
+		endpoint = methodAttrs.endpoint
+		httpVerb = methodAttrs.verb
+		// If both these are empty string, RPC is not allowed for this method
+		if endpoint != "" || httpVerb != "" {
+			if !strings.HasPrefix(string(endpoint), "/") {
+				regFail("%s: endpoint URL must start with /, got %q", lowerName, endpoint)
+			}
+			if httpVerb != http.MethodGet && httpVerb != http.MethodPost && httpVerb != http.MethodPut {
+				regFail("%s: unknown http verb, got %q", lowerName, httpVerb)
+			}
+		}
+
 		// Save the method to the registration table
 		reg[funcName] = callable{
 			Impl:      impl,
 			Func:      i.Func,
 			InType:    inType,
 			OutType:   outType,
 			RetCursor: returnsCursor,
+			Endpoint:  endpoint,
+			Verb:      httpVerb,
 		}
 		log.Debugf("%d: registered %s(*%s) %v", k, funcName, inType, outType)
 	}
 
 	for k := range methodMap {
-		if k != "Name" {
-			panic(fmt.Sprintf("%s: did not find implementation for method %s", msetType, k))
+		if k != "Name" && k != "Attributes" {
```

Comment:
I think this check means that any `MethodSet` can only have "method functions" and the `MethodSet` interface implementation (formally: `Name` and `Attributes` methods). Can we get a comment on the `MethodSet` interface definition:

```
MethodSet implementations cannot have extra methods, only MethodSet interface methods and methods to be registered
```

Reply:
Added clarifying comment